subprojects/soak/src/integTest/groovy/org/gradle/vfs/VirtualFileSystemRetentionSoakTest.groovy

```diff
@@ -0,0 +1,150 @@
+
+@Category(SoakTest)
+class VirtualFileSystemRetentionSoakTest extends DaemonIntegrationSpec {
+
+    private static final int NUMBER_OF_SUBPROJECTS = 50
+    private static final int NUMBER_OF_SOURCES_PER_SUBPROJECT = 100
+    private static final int TOTAL_NUMBER_OF_SOURCES = NUMBER_OF_SUBPROJECTS * NUMBER_OF_SOURCES_PER_SUBPROJECT
+    private static final double LOST_EVENTS_RATIO_MAC_OS = 0.75
+    private static final double LOST_EVENTS_RATIO_WINDOWS = 0.1
+
+    List<TestFile> sourceFiles
+
+    def setup() {
+        // Use 20 minutes idle timeout since the test may be running longer with an idle daemon
+        executer.withDaemonIdleTimeoutSecs(1200)
+        def subprojects = (1..NUMBER_OF_SUBPROJECTS).collect { "project$it" }
+        def rootProject = multiProjectBuild("javaProject", subprojects) {
+            buildFile << """
+                allprojects {
+                    apply plugin: 'java-library'
+
+                    tasks.withType(JavaCompile).configureEach {
+                        options.fork = true
+                    }
+                }
+            """
+        }
+        sourceFiles = subprojects.collectMany { projectDir ->
+            (1..NUMBER_OF_SOURCES_PER_SUBPROJECT).collect {
+                def sourceFile = rootProject.file("${projectDir}/src/main/java/my/domain/Dummy${it}.java")
+                modifySourceFile(sourceFile, 0)
+                return sourceFile
+            }
+        }
+
+        executer.beforeExecute {
+            withArgument("-D${VirtualFileSystemServices.VFS_RETENTION_ENABLED_PROPERTY}=true")
+        }
+    }
+
+    def "file watching works with multiple builds on the same daemon"() {
+        when:
+        succeeds("assemble")
+        def daemon = daemons.daemon
+        then:
+        daemon.assertIdle()
+
+        expect:
+        (1..100).each { iteration ->
+            changeAllSourceFiles(iteration)
+            waitForChangesToBePickedUp()
+            succeeds("assemble", "--parallel")
+            assert daemons.daemon.logFile == daemon.logFile
+            daemon.assertIdle()
+            assertWatchingSucceeded()
+            assert receivedFileSystemEvents >= expectedFileSystemEvents(TOTAL_NUMBER_OF_SOURCES, 1)
+        }
+    }
+
+    def "file watching works with many changes between two builds"() {
+        when:
+        succeeds("assemble", "--parallel")
+        def daemon = daemons.daemon
+        then:
+        daemon.assertIdle()
+
+        when:
+        (1..100).each { iteration ->
+            changeAllSourceFiles(iteration)
+            waitForChangesToBePickedUp()
+        }
+        then:
+        succeeds("assemble")
+        daemons.daemon.logFile == daemon.logFile
+        daemon.assertIdle()
+        assertWatchingSucceeded()
+        receivedFileSystemEvents >= expectedFileSystemEvents(TOTAL_NUMBER_OF_SOURCES, 100)
+    }
+
+    private static int expectedFileSystemEvents(int numberOfChangedFiles, int numberOfChangesPerFile) {
```

Comment:
This should be called `minimumExpectedFileSystemEvents()`.

Better yet we could have a method called `assertEventCount()` so we can check an upper and lower limit or something.

Reply:
I went for `minimumExpectedFileSystemEvents`. See https://github.com/gradle/gradle/issues/12276 for improving on that.